source/grid/grid_tools.cc

```diff
@@ -872,35 +883,55 @@ namespace GridTools
     std::size_t n_negative_cells = 0;
     for (auto &cell : cells)
       {
-        Assert(cell.vertices.size() ==
-                 ReferenceCells::get_hypercube<dim>().n_vertices(),
-               ExcNotImplemented());
         const ArrayView<const unsigned int> vertices(cell.vertices);
-        if (GridTools::cell_measure(all_vertices, vertices) < 0)
+        if (cell.vertices.size() ==
+            ReferenceCells::get_hypercube<dim>().n_vertices())
           {
-            ++n_negative_cells;
-
-            // TODO: this only works for quads and hexes
-            if (dim == 2)
+            if (GridTools::cell_measure(all_vertices, vertices) < 0)
```

Comment:
I would pull the call to `cell_measure()` out front, and only then when this call return true have the `if` statements about the reference cell.

You can write the call to figure out which reference cell you have in an easier to read way if you use `if (ReferenceCell::n_vertices_to_type(dim, vertices.size()).is_hyper_cube())` and similarly for `...is_simplex()`.